### PR TITLE
LWDEV-10481 Implement mail notification in trade volume service

### PR DESCRIPTION
**Pull request by olpapchenko**

### CODE REVIEW

File: src/dist/cfg/log4j2.xml [hunk 1]
```diff
@@ -39,6 +51,10 @@
             <AppenderRef ref="IncomingEventProcessLog"/>
         </AsyncLogger>
 
+        <Logger name="com.lykke.trade.volume.monitoring.service.notification.impl.NotificationServiceImpl" level="info">
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: additivity=false ?

File: build.gradle [hunk 2]
```diff
@@ -32,23 +32,26 @@ repositories {
...
+        outputs.dir(swaggerOutputDir)
+        doLast {
+            def config = new CodegenConfigurator()
+            config.setInputSpec(input)
+            config.setOutputDir(swaggerOutputDir.path)
+            config.setLang('java')
+            config.setAdditionalProperties([
+                    'invokerPackage': 'com.lykke.trade.volume.monitoring.service.loader.http.generated.client',
+                    'modelPackage'  : 'com.lykke.trade.volume.monitoring.service.loader.http.generated.client.model',
+                    'apiPackage'    : 'com.lykke.trade.volume.monitoring.service.loader.http.generated.client.api',
```
Comment (barsukovskii) [inline build.gradle]: If we use the same package for both api it is better to rename it as `mailNotificationApi` is not `loader`

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/config/NotificationsConfig.kt [hunk 3]
```diff
@@ -1,3 +1,11 @@
...
-class NotificationsConfig (val throttlingPeriod: Long)
\ No newline at end of file
+import com.lykke.trade.volume.monitoring.service.entity.MailApiType
+
+class NotificationsConfig(
+        val type: MailApiType,
+        val azureConfig: AzureNotificationConfig?,
+        val httpConfig: HttpApiConfig?,
+        val throttlingPeriod: Long,
+        val senderAddress: String,
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/config/NotificationsConfig.kt]: This field is used only in AzureSender. Does HttpApiSender have its own non configurable senderAddress? If yes lets switch this field type to nullable

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/spring/MailConfig.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Autowired
+    private lateinit var config: Config
+
+    @Autowired
+    private lateinit var mailValidator: MailValidator
+
+    @Bean
+    fun mailNotificationService(
+            @Value("\${mail.notification.format}")
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/spring/MailConfig.kt]: Lets add the prefix `azure` to the property name and make the parameter nullable

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/impl/NotificationServiceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                            .setTargetAssetId(targetAssetId)
+                            .setAssetId(request.assetId)
+                            .setTimestamp(request.timestamp)
+                            .build()
+
+                    mailNotificationService.sendMail(notificationsConfig.mailAddresses, messageSubject, message)
+                    sentNotificationsCache.add(clientId = request.clientId, assetId = request.assetId)
+                    LOGGER.info(message)
+                } catch (e: Exception) {
+                    LOGGER.error("Error occurred when sending mail notification - volume limit reached for client ${request.clientId}, assetId: ${request.assetId}")
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/impl/NotificationServiceImpl.kt]: Propose to add stacktrace to log

File: src/dist/cfg/log4j2.xml (hunk 1 above)
Comment (olpapchenko) [reply src/dist/cfg/log4j2.xml]: fixed

File: build.gradle (hunk 2 above)
Comment (olpapchenko) [reply build.gradle]: removed `loader` from the package name

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/config/NotificationsConfig.kt (hunk 3 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/config/NotificationsConfig.kt]: fixed

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/spring/MailConfig.kt (hunk 4 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/spring/MailConfig.kt]: fixed

File: src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/impl/NotificationServiceImpl.kt (hunk 5 above)
Comment (olpapchenko) [reply src/main/kotlin/com/lykke/trade/volume/monitoring/service/notification/impl/NotificationServiceImpl.kt]: fixed

**Review by barsukovskii — APPROVED** [review]

**Pull request merged**